classes/kohana/encrypt.php

```diff
@@ -1,213 +1,121 @@
  * @author     Kohana Team
  * @copyright  (c) 2007-2012 Kohana Team
  * @license    http://kohanaframework.org/license
  */
 class Kohana_Encrypt {
-
 	/**
-	 * @var  string  default instance name
+	 * @var string
 	 */
-	public static $default = 'default';
+	private static $default = 'default';
 
 	/**
-	 * @var  array  Encrypt class instances
+	 * @var array
 	 */
-	public static $instances = array();
+	private static $instances = [];
 
 	/**
-	 * @var  string  OS-dependent RAND type to use
+	 * @var string
 	 */
-	protected static $_rand;
+	private $key;
 
 	/**
-	 * Returns a singleton instance of Encrypt. An encryption key must be
-	 * provided in your "encrypt" configuration file.
-	 *
-	 *     $encrypt = Encrypt::instance();
-	 *
-	 * @param   string  $name   configuration group name
-	 * @return  Encrypt
+	 * @param string $key Encryption key (hexadecimal)
 	 */
-	public static function instance($name = NULL)
+	public function __construct($key)
 	{
-		if ($name === NULL)
-		{
-			// Use the default instance name
-			$name = Encrypt::$default;
-		}
-
-		if ( ! isset(Encrypt::$instances[$name]))
-		{
-			// Load the configuration data
-			$config = Kohana::$config->load('encrypt')->$name;
-
-			if ( ! isset($config['key']))
-			{
-				// No default encryption key is provided!
-				throw new Kohana_Exception('No encryption key is defined in the encryption configuration group: :group',
-					array(':group' => $name));
-			}
-
-			if ( ! isset($config['mode']))
-			{
-				// Add the default mode
-				$config['mode'] = MCRYPT_MODE_NOFB;
-			}
-
-			if ( ! isset($config['cipher']))
-			{
-				// Add the default cipher
-				$config['cipher'] = MCRYPT_RIJNDAEL_128;
-			}
-
-			// Create a new instance
-			Encrypt::$instances[$name] = new Encrypt($config['key'], $config['mode'], $config['cipher']);
-		}
-
-		return Encrypt::$instances[$name];
+		$this->key = hex2bin($key);
 	}
 
 	/**
-	 * Creates a new mcrypt wrapper.
+	 * Encrypts the data and returns it as a base 64 encoded string.
 	 *
-	 * @param   string  $key    encryption key
-	 * @param   string  $mode   mcrypt mode
-	 * @param   string  $cipher mcrypt cipher
+	 * @param string $data
+	 *
+	 * @return string
 	 */
-	public function __construct($key, $mode, $cipher)
+	public function encode($data)
 	{
-		// Find the max length of the key, based on cipher and mode
-		$size = mcrypt_get_key_size($cipher, $mode);
-
-		if (isset($key[$size]))
-		{
-			// Shorten the key to the maximum size
-			$key = substr($key, 0, $size);
-		}
+		$nonce = random_bytes(SODIUM_CRYPTO_SECRETBOX_NONCEBYTES);
+		$encryptedData = sodium_crypto_secretbox($data, $nonce, $this->key);
 
-		// Store the key, mode, and cipher
-		$this->_key    = $key;
-		$this->_mode   = $mode;
-		$this->_cipher = $cipher;
-
-		// Store the IV size
-		$this->_iv_size = mcrypt_get_iv_size($this->_cipher, $this->_mode);
+		// Base 64 encode the encrypted data.
+		return base64_encode(bin2hex($nonce.$encryptedData));
```

Comment:
Not very efficient converting to hex then base64 encoding it, I don't see any reason to add the additional hex conversion step.